File: src/main/java/com/chatdemo/ModelFactory.java
package com.chatdemo;

import dev.langchain4j.model.chat.ChatModel;
import dev.langchain4j.model.chat.StreamingChatModel;
import dev.langchain4j.model.anthropic.AnthropicChatModel;
import dev.langchain4j.model.anthropic.AnthropicStreamingChatModel;
import dev.langchain4j.model.googleai.GoogleAiGeminiChatModel;
import dev.langchain4j.model.googleai.GoogleAiGeminiStreamingChatModel;
import dev.langchain4j.model.image.ImageModel;
import dev.langchain4j.model.openai.OpenAiChatModel;
import dev.langchain4j.model.openai.OpenAiImageModel;
import dev.langchain4j.model.openai.OpenAiStreamingChatModel;

/**
 * Factory for creating LangChain4j ChatModel instances.
 */
public class ModelFactory {
    
    private static final String GROK_BASE_URL = "https://api.x.ai/v1";
    private static final String OPENAI_IMAGE_MODEL = "chatgpt-image-latest";
    private static final String GEMINI_IMAGE_MODEL = "gemini-2.5-flash-image";
    private static final String GROK_IMAGE_MODEL = "grok-imagine-image";
    
    /**
     * Create a ChatModel based on the provider configuration.
     */
    public static ChatModel createModel(ProviderConfig config) {
        return switch (config.providerType().toLowerCase()) {
            case "gemini" -> createGeminiModel(config);
            case "chatgpt" -> createOpenAIModel(config);
            case "claude" -> createClaudeModel(config);
            case "grok" -> createGrokModel(config);
            default -> throw new IllegalArgumentException("Unknown provider: " + config.providerType());
        };
    }

    /**
     * Create a StreamingChatLanguageModel based on the provider configuration.
     */
    public static StreamingChatModel createStreamingModel(ProviderConfig config) {
        return switch (config.providerType().toLowerCase()) {
            case "gemini" -> createGeminiStreamingModel(config);
            case "chatgpt" -> createOpenAiStreamingModel(config);
            case "claude" -> createClaudeStreamingModel(config);
            case "grok" -> createGrokStreamingModel(config);
            default -> throw new IllegalArgumentException("Unknown provider: " + config.providerType());
        };
    }

    /**
     * Create a Gemini ChatModel for image generation outputs.
     */
    public static ChatModel createGeminiImageModel(String apiKey, String modelName) {
        return GoogleAiGeminiChatModel.builder()
            .apiKey(apiKey)
            .modelName(modelName)
            .build();
    }

    /**
     * Create an ImageModel for OpenAI image generation.
     */
    public static ImageModel createImageModel(String apiKey) {
        return createImageModel(apiKey, OPENAI_IMAGE_MODEL);
    }

    public static ImageModel createImageModel(String apiKey, String modelName) {
        return OpenAiImageModel.builder()
            .apiKey(apiKey)
            .modelName(modelName)
            .responseFormat("url")
            .build();
    }

    public static ImageModel createGrokImageModel(String apiKey, String modelName) {
        return OpenAiImageModel.builder()
            .apiKey(apiKey)
            .modelName(modelName)
            .baseUrl(GROK_BASE_URL)
            .build();
    }

    public static String defaultOpenAiImageModel() {
        return OPENAI_IMAGE_MODEL;
    }

    public static String defaultGeminiImageModel() {
        return GEMINI_IMAGE_MODEL;
    }

    public static String defaultGrokImageModel() {
        return GROK_IMAGE_MODEL;
    }
    
    private static ChatModel createGeminiModel(ProviderConfig config) {
        return GoogleAiGeminiChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .build();
    }

    private static StreamingChatModel createGeminiStreamingModel(ProviderConfig config) {
        return GoogleAiGeminiStreamingChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .build();
    }
    
    private static ChatModel createOpenAIModel(ProviderConfig config) {
        return OpenAiChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .build();
    }

    private static StreamingChatModel createOpenAiStreamingModel(ProviderConfig config) {
        return OpenAiStreamingChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .build();
    }
    
    private static ChatModel createClaudeModel(ProviderConfig config) {
        return AnthropicChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .build();
    }

    private static StreamingChatModel createClaudeStreamingModel(ProviderConfig config) {
        return AnthropicStreamingChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .build();
    }
    
    private static ChatModel createGrokModel(ProviderConfig config) {
        // Grok uses OpenAI-compatible API with custom base URL
        return OpenAiChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .baseUrl(GROK_BASE_URL)
            .build();
    }

    private static StreamingChatModel createGrokStreamingModel(ProviderConfig config) {
        // Grok uses OpenAI-compatible API with custom base URL
        return OpenAiStreamingChatModel.builder()
            .apiKey(config.apiKey())
            .modelName(config.model())
            .baseUrl(GROK_BASE_URL)
            .build();
    }
}

File: src/main/scala/com/chatdemo/backend/model/ModelFactory.scala
package com.chatdemo.backend.model

import com.chatdemo.common.config.ProviderConfig
import dev.langchain4j.model.chat.{ChatModel, StreamingChatModel}
import dev.langchain4j.model.anthropic.{AnthropicChatModel, AnthropicStreamingChatModel}
import dev.langchain4j.model.googleai.{GoogleAiGeminiChatModel, GoogleAiGeminiStreamingChatModel}
import dev.langchain4j.model.image.ImageModel
import dev.langchain4j.model.openai.{OpenAiChatModel, OpenAiImageModel, OpenAiStreamingChatModel}

/**
 * Factory for creating LangChain4j ChatModel instances.
 */
object ModelFactory {

  private val GrokBaseUrl = "https://api.x.ai/v1"
  private val QwenApiBaseUrl = "https://dashscope-intl.aliyuncs.com/api/v1"
  private val OpenAiImageModelName = "dall-e-3"
  private val GeminiImageModelName = "gemini-2.5-flash-image"
  private val GrokImageModelName = "grok-imagine-image"
  private val QwenCreateImageModelName = "qwen-image-max"
  private val QwenEditImageModelName = "qwen-image-edit-max"

  /** Create a ChatModel based on the provider configuration. */
  def createModel(config: ProviderConfig): ChatModel = {
    config.providerType.toLowerCase match {
      case "gemini"  => createGeminiModel(config)
      case "chatgpt" => createOpenAIModel(config)
      case "claude"  => createClaudeModel(config)
      case "grok"    => createGrokModel(config)
      case _         => throw new IllegalArgumentException("Unknown provider: " + config.providerType)
    }
  }

  /** Create a StreamingChatModel based on the provider configuration. */
  def createStreamingModel(config: ProviderConfig): StreamingChatModel = {
    config.providerType.toLowerCase match {
      case "gemini"  => createGeminiStreamingModel(config)
      case "chatgpt" => createOpenAiStreamingModel(config)
      case "claude"  => createClaudeStreamingModel(config)
      case "grok"    => createGrokStreamingModel(config)
      case _         => throw new IllegalArgumentException("Unknown provider: " + config.providerType)
    }
  }

  /** Create a Gemini ChatModel for image generation outputs. */
  def createGeminiImageModel(apiKey: String, modelName: String): ChatModel = {
    GoogleAiGeminiChatModel.builder()
      .apiKey(apiKey)
      .modelName(modelName)
      .build()
  }

  /** Create an ImageModel for OpenAI image generation. */
  def createImageModel(apiKey: String): ImageModel = {
    createImageModel(apiKey, OpenAiImageModelName)
  }

  def createImageModel(apiKey: String, modelName: String): ImageModel = {
    OpenAiImageModel.builder()
      .apiKey(apiKey)
      .modelName(modelName)
      .responseFormat("url")
      .build()
  }

  def createGrokImageModel(apiKey: String, modelName: String): ImageModel = {
    OpenAiImageModel.builder()
      .apiKey(apiKey)
      .modelName(modelName)
      .baseUrl(GrokBaseUrl)
      .build()
  }

  def createQwenImageModel(apiKey: String, modelName: String): ImageModel = {
    OpenAiImageModel.builder()
      .apiKey(apiKey)
      .modelName(modelName)
      .baseUrl("https://dashscope-intl.aliyuncs.com/compatible-mode/v1")
      .responseFormat("url")
      .build()
  }

  def defaultOpenAiImageModel: String = OpenAiImageModelName
  def defaultGeminiImageModel: String = GeminiImageModelName
  def defaultGrokImageModel: String = GrokImageModelName
  def defaultQwenCreateImageModel: String = QwenCreateImageModelName
  def defaultQwenEditImageModel: String = QwenEditImageModelName
  def qwenApiBaseUrl: String = QwenApiBaseUrl

  private def createGeminiModel(config: ProviderConfig): ChatModel = {
    GoogleAiGeminiChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .build()
  }

  private def createGeminiStreamingModel(config: ProviderConfig): StreamingChatModel = {
    GoogleAiGeminiStreamingChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .build()
  }

  private def createOpenAIModel(config: ProviderConfig): ChatModel = {
    OpenAiChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .build()
  }

  private def createOpenAiStreamingModel(config: ProviderConfig): StreamingChatModel = {
    OpenAiStreamingChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .build()
  }

  private def createClaudeModel(config: ProviderConfig): ChatModel = {
    AnthropicChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .build()
  }

  private def createClaudeStreamingModel(config: ProviderConfig): StreamingChatModel = {
    AnthropicStreamingChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .build()
  }

  private def createGrokModel(config: ProviderConfig): ChatModel = {
    OpenAiChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .baseUrl(GrokBaseUrl)
      .build()
  }

  private def createGrokStreamingModel(config: ProviderConfig): StreamingChatModel = {
    OpenAiStreamingChatModel.builder()
      .apiKey(config.apiKey)
      .modelName(config.model)
      .baseUrl(GrokBaseUrl)
      .build()
  }
}
